pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -27,65 +29,103 @@ import (
+
+	podResources    v1.ResourceList
+	daemonResources v1.ResourceList
 }
 
-func NewNode(constraints *v1alpha5.Constraints, instanceTypeOptions []cloudprovider.InstanceType, pods ...*v1.Pod) *Node {
-	n := &Node{Constraints: constraints.DeepCopy()}
+func NewNode(constraints *v1alpha5.Constraints, daemonResources v1.ResourceList, instanceTypeOptions []cloudprovider.InstanceType) *Node {
+	n := &Node{
+		Constraints:     *constraints.DeepCopy(),
+		daemonResources: daemonResources,
+	}
 
 	for _, it := range instanceTypeOptions {
+		// If a zero-resource pod can't fit, don't consider this instance type.  This occurs if the node overhead +
+		// daemon set requests is already larger than the instance type can support
+		if !n.newPodCanFit(nil, it) {
+			continue
+		}
+
 		// pre-filter our list of all possible instance types by what the provisioner allows
 		if !cloudprovider.Compatible(it, constraints.Requirements) {
 			continue
 		}
 		n.InstanceTypeOptions = append(n.InstanceTypeOptions, it)
 	}
-	// technically we might create some invalid construct here if the pod can't be supported by any instance type
-	// but this is ok as we just won't have any valid instance types, this pod won't be schedulable and nothing else
-	// will be compatible with this node
-	for _, p := range pods {
-		n.Add(p)
-	}
 	return n
 }
 
 func (n Node) Compatible(pod *v1.Pod) error {
-	if n.unschedulable {
-		return errors.New("node is unschedulable")
-	}
-
 	podRequirements := v1alpha5.NewPodRequirements(pod)
-	if err := n.Constraints.Requirements.Compatible(podRequirements); err != nil {
+	if err := n.Requirements.Compatible(podRequirements); err != nil {
 		return err
 	}
 
-	tightened := n.Constraints.Requirements.Add(podRequirements.Requirements...)
+	tightened := n.Requirements.Add(podRequirements.Requirements...)
 	// Ensure that at least one instance type of the instance types that we are already narrowed down to based on the
 	// existing pods can support the pod resources and combined pod + provider requirements
+	podRequests := resources.RequestsForPods(pod)
 	for _, it := range n.InstanceTypeOptions {
-		if cloudprovider.Compatible(it, tightened) && n.hasCompatibleResources(resources.RequestsForPods(pod), it) {
+		newSize := resources.Merge(n.reservedResources(it), podRequests)
+		if cloudprovider.Compatible(it, tightened) &&
+			n.newPodCanFit(newSize, it) &&
+			n.hasCompatibleResources(podRequests, it) {
 			return nil
 		}
 	}
 	return errors.New("no matching instance type found")
 }
 
+func (n Node) reservedResources(it cloudprovider.InstanceType) v1.ResourceList {
+	return resources.Merge(it.Overhead(), n.daemonResources, n.podResources)
+}
+
+func (n *Node) newPodCanFit(newSize v1.ResourceList, it cloudprovider.InstanceType) bool {
+	for resourceName, totalQuantity := range it.Resources() {
+		reservedQuantity := newSize[resourceName]
+		if reservedQuantity.Cmp(totalQuantity) > 0 {
```

Comment:
since we made a helper, can we use it to inline this? `resource.Cmp(newSize[resourceName], totalQuantity) > 0`